### DISCUSSION TITLE
ParsedMaterial

### DISCUSSION THREAD

**Discussion by selarem** [idaholab/moose#24758] [Q&A Modules: Phase field]
Hi moose developpers,
we have implemented a material model that needs to input a  vector variable. Everything is working as expected.
To apply the model for more complex situations, we want the vector  a function of (x,y,z, ... and includes random aspects).
I tried using ParsedMaterial. 
```
[./define_x]
   type = ParsedMaterial
   args= 'x y z'
   f_name = 'Xvect'
   function = 'x + y +z'
  [../]
```
and wrote :

plane_normal = 'Xvect Xvect/2  Xvect/3.0'

cannot convert field 'Materials/Stress/plane_normal' value 'Xvect' to float.

Thanks for any help.

Sab

**Comment by GiudGiud** [comment]
Hello

you are trying to define a vector material property right?

if so you ll need to proceed component by component, or create a parsedVectorMaterial

Guillaume

**Reply by ghost** [reply]
Hello, 

Thank you for your reply. 

Even when using the ParsedVectorFunction the error remains the same "**cannot convert field 'Kernels/ACInterfaceCleavageFracture/cleavage_plane_normal' value 'vectorexp' to float**" 

The input file Function bloc looks as follows : 
```
[Functions]
  [vectorexp]
    type = ParsedVectorFunction
    expression_x = 'x'
    expression_y = 'y/2'
    expression_z = '0'
  [../]
[]
```
And the Kernel :
```
  [ACInterfaceCleavageFracture]
    type = ACInterfaceCleavageFracture
    variable = c
    beta_penalty = 1
    cleavage_plane_normal = vectorexp
  [../]
```

**Reply by ghost** [reply]
Hello, 

The issue is resolved, we managed to fix the problem and the kernel can now read the parsed vector function coordinates. Though we now have another question : while making sure that the coordinates are different for each element of the mesh (by printing out the x and y coordinates) we only got the values for one element, so only 4 values (we assume these are values for each qp of the element) even when our mesh is composed of 2 or 3 or more elements. The question is why does it only print out values for one of the elements, and why is it the last element (almost always). 

On a related note, I have a question regarding the SetupQuadratureAction (I am attempting to have one quadrature point for each element instead of four), I am unsure about the distinction between the 'order' and 'element order' input parameters  and if you could provide an explanation, it would be greatly appreciated.

Thank you :) 

Nour

**Reply by GiudGiud** [reply]
order is the general one, then element_order and side_order are for differentiating

**Reply by GiudGiud** [reply]
so you added prints inside your material and only saw one element?
Are you using the software in parallel?
Is the material block restricted? Is it constant per element?

**Reply by ghost** [reply]
This is what the input file looks like with a 3 element mesh: 

```
[Mesh]
  [mesh]
    type = GeneratedMeshGenerator
    dim = 2
    xmax =3
    nx = 3
    ny = 1
    elem_type = quad4
[]
[]

[Functions]
  [vectorexp]
    type = ParsedVectorFunction
    expression_x = 'x'
    expression_y = 'y'
    expression_z = '0'
  [../]
[]

[Kernels]
[./ACInterfaceCleavageFracture]
    type = meACInterfaceCleavageFracture
    variable = c
    beta_penalty = 1
    cleavage_plane_normal = vectorexp
  [../]
```

with the meACInterfaceCleavageFracture including the print option here's the excerpt from this file : 

```
meACInterfaceCleavageFracture::meACInterfaceCleavageFracture(const InputParameters & parameters)
  : ACInterface(parameters),
    _beta_penalty(getParam<Real>("beta_penalty")),
   _cleavage_plane_normal2(getFunction("cleavage_plane_normal"))
{
}

Real
meACInterfaceCleavageFracture::betaNablaPsi()
{

libMesh::VectorValue<Real> _cleavage_plane_normal ;
_cleavage_plane_normal(0) = _cleavage_plane_normal2.vectorValue(_t, _q_point[_qp])(0);
_cleavage_plane_normal(1) = _cleavage_plane_normal2.vectorValue(_t, _q_point[_qp])(1);
_cleavage_plane_normal(2) = _cleavage_plane_normal2.vectorValue(_t, _q_point[_qp])(2);

std::cout << "from K"<< _cleavage_plane_normal(0) << " " << _cleavage_plane_normal(1) << " " << _cleavage_plane_normal(2)<< std::endl ;

```

and here's the output where we have a repetitive 4 points coordinates printout : 

```
Time Step 1, time = 5e-05, dt = 5e-05
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
from K2.21132 0.211325 0
from K2.78868 0.211325 0
from K2.21132 0.788675 0
from K2.78868 0.788675 0
 0 Nonlinear |R| = ^[[32m0.000000e+00^[[39m
^[[32m Solve Converged!^[[39m
```

**Reply by ghost** [reply]
Our main concern is to check whether we get distinct cleavage plane normals for each qp with this syntax or if it's the same cleavage plane normal for all qps.

**Reply by ghost** [reply]
Problem solved, we are now able to input a different vector for each gaussian point. 
We now would like to be able to input random vector components (using ParsedVectorFunction or any equivalent syntax in the input file). In fact, we want this vector to remain unchanged during the entire computation, for each qp.

Regards,

Nour

**Reply by GiudGiud** [reply]
Hello

You ll want to use the RandomIC on a variable then use that for the random number
If it s not quite what you want and you cant find a conversion from the variable to what you need you can make a new object in the category you need and use the RandomInterface

Guillaume

**Reply by ghost** [reply]
thank you Guillaume for your reply,
I thought that RandomIC is for initial conditions like initial temperature or damage, ... to initiate variables we are solving for.
But what we want is something like this
```
[Functions]
  [vectorexp]
    type = ParsedVectorFunction
    expression_x = 'sin(x)*cos(y)'
    expression_y = 'cos(x)*sin(2*y)'
    expression_z = '0'
  [../]
[]
```
with some random instead of sine or cosine functions. By the way I wanted to know the list of availble functions.
I know that in C++ we could have random using the standard library, but how to have this possible in input files ?

regards,

**Reply by GiudGiud** [reply]
hello

no you can't call C++ directly from the input file sadly.
I was thinking the randomIC would give you a field of random values and you could use that later on.
In a function, you would not be able to use a variable though. Only in another auxvariable or a material could you use the value of a variable.

It seems you will have to code your own object, or look at adding rand() in the operations that the Parsed expressions support

Guillaume

**Reply by ghost** [reply]
Hello, 

Ok, thank you. So to create our own object, we'd like to have a list of the functions available to use in the input file for the Parsed vector function (just like cosine and sine are available, what are the other available functions). Thank you.

Nour

**Reply by GiudGiud** [reply]
You ll want to look at the documentation for fparser:
http://warp.povusers.org/FunctionParser/fparser.html

towards the bottom is the list of operations that are supported
actually it looks like they have an example for random numbers

**Reply by ghost** [reply]
We want just to be able to use rand() in the input file as we use classical function (sin, cos, abs, tan ...). We could not find the source file in moose that makes possible the use of the classical function that we could add the rand() function. 
regards,

**Reply by GiudGiud** [reply]
I dont know that this is going to be possible. Looking in fparser code, I dont see rand. 
You ll want to work on libmesh to enable this, we inherit this capability from libmesh pulling in fparser

**Reply by GiudGiud** [reply]
tbh it would be easier to code a new material with the expression you need

**Comment by selarem** [comment]
If coded inside a material, the random property will change at every time
step which is not what we want. We want something like  grains orientation
in CP that could be randomly fixed at the beginning of the computation and
remains unchanged after.
Regards

**Reply by GiudGiud** [reply]
no you can fix a material property to its initial value
set `compute = false` and compute it only once

**Comment by selarem** [comment]
Normally it should not be that complicated. We have been trying with
initQpStatefulProperties,  initialSetup and the class constructor and
obtained undesirable outputs. I wonder what the initialization methods are
used for and when are they called. Example when (in
initQpStatefulProperties) I use _q_point[_qp] in a vector components that I
want to vary for each qp, I obtain the same vector for all elements which
is the one corresponding to the last qp.
I could provide more example to show that it is not clear for me how things
work and interact in moose.
Regards,

**Reply by GiudGiud** [reply]
I raised an issue.

I m not sure you want a stateful material?
Stateful material are meant to be able to retrieve an old value from a material property, to build semi-explicit schemes (explicit in some properties)

Dont you just want a "constant" property? Initialized once with a random field and that's it?

**Comment by selarem** [comment]
We would like to have  a RealVectorValue which is randomly generated for
each _qp. Once generated it remains constant during the computation.
Like if we generate a Young modulus  or a Poisson ratio that depends on the
_qp initial coordinates.
regards,

**Reply by GiudGiud** [reply]
I think this was mentioned already but this should work:

Look at the time step in the material, and only compute the material if the current time step is equal to the first time step

**Comment by selarem** [comment]
I have tried this, it generates the same random vector  for all _qp. What we want is one different vector for each _qp.
if (_t == _dt){
         std::cout <<"Element  = :" <<  _current_elem->id() << std::endl;
  RealVectorValue cp;
 cp(0) = MooseRandom::rand();
 cp(1) = MooseRandom::rand();
 cp(2) = 0.0*MooseRandom::rand();

  std::cout << " VectorX0 "<< cp(0) <<" VectY0 "<< cp(1) <<" VectZ0 " << cp(2) << " norm0 " << cp.norm()  << std::endl;

 Real normeCP = std::sqrt(std::pow(cp(0),2) + std::pow(cp(1),2) + std::pow(cp(2),2) ) ;

 cp(0)/=normeCP;
 cp(1)/=normeCP;
 cp(2)/=normeCP;

  std::cout << " VectorX  = "<< cp(0) <<" VectY =  "<< cp(1) <<" VectZ =  " << cp(2) << " norm =  " << cp.norm()  << std::endl;

 _rand_cleavage_plane_normal[_qp] = cp;
}

**When using one single element (_qp):** 

Time Step 1, time = 5e-08, dt = 5e-08
Element  = :0
 VectorX0 0.548814 VectY0 0.715189 VectZ0 0 norm0 0.901494
 VectorX  = 0.608782 VectY =  0.793338 VectZ =  0 norm =  1
 *******
 VectorX 0.608782 VectY 0.793338 VectZ 0 norm 1
Element  = :0
 VectorX0 0.544883 VectY0 0.423655 VectZ0 0 norm0 0.690204
 VectorX  = 0.789453 VectY =  0.613811 VectZ =  0 norm =  1
 *******
 VectorX 0.789453 VectY 0.613811 VectZ 0 norm 1
Element  = :0
 VectorX0 0.437587 VectY0 0.891773 VectZ0 0 norm0 0.993349
 VectorX  = 0.440517 VectY =  0.897744 VectZ =  0 norm =  1
 *******
 VectorX 0.440517 VectY 0.897744 VectZ 0 norm 1
Element  = :0
 VectorX0 0.383442 VectY0 0.791725 VectZ0 0 norm0 0.879691
 VectorX  = 0.435882 VectY =  0.900004 VectZ =  0 norm =  1
 *******
 VectorX 0.435882 VectY 0.900004 VectZ 0 norm 1
Element  = :0
 VectorX0 0.568045 VectY0 0.925597 VectZ0 0 norm0 1.086
 VectorX  = 0.52306 VectY =  0.852296 VectZ =  0 norm =  1
 *******
 VectorX 0.52306 VectY 0.852296 VectZ 0 norm 1
^[[32m Solve Converged!^[[39m
Element  = :0
 VectorX0 0.0871293 VectY0 0.0202184 VectZ0 0 norm0 0.0894444
 VectorX  = 0.974117 VectY =  0.226044 VectZ =  0 norm =  1
 *******
 VectorX 0.974117 VectY 0.226044 VectZ 0 norm 1

Time Step 2, time = 1e-07, dt = 5e-08
Element  = :0
 *******
 VectorX 0.974117 VectY 0.226044 VectZ 0 norm 1
Element  = :0
 *******
 VectorX 0.974117 VectY 0.226044 VectZ 0 norm 1
Element  = :0
 *******

**When unsing  2 Elements:**

Element  = :0
 VectorX0 0.118274 VectY0 0.639921 VectZ0 0 norm0 0.650759
 VectorX  = 0.181748 VectY =  0.983345 VectZ =  0 norm =  1
 *******
 VectorX 0.181748 VectY 0.983345 VectZ 0 norm 1
Element  = :1
 VectorX0 0.944669 VectY0 0.521848 VectZ0 0 norm0 1.07922
 VectorX  = 0.875322 VectY =  0.48354 VectZ =  0 norm =  1
 *******
 VectorX 0.875322 VectY 0.48354 VectZ 0 norm 1
^[[32m Solve Converged!^[[39m
Element  = :0
 VectorX0 0.264556 VectY0 0.774234 VectZ0 0 norm0 0.818185
 VectorX  = 0.323344 VectY =  0.946281 VectZ =  0 norm =  1
 *******
 VectorX 0.323344 VectY 0.946281 VectZ 0 norm 1
Element  = :1
 VectorX0 0.568434 VectY0 0.0187898 VectZ0 0 norm0 0.568744
 VectorX  = 0.999454 VectY =  0.0330373 VectZ =  0 norm =  1
 *******
 VectorX 0.999454 VectY 0.0330373 VectZ 0 norm 1

**Time Step 2, time = 1e-07, dt = 5e-08**
Element  = :0
 *******
 VectorX 0.999454 VectY 0.0330373 VectZ 0 norm 1
Element  = :1
 *******
 VectorX 0.999454 VectY 0.0330373 VectZ 0 norm 1
Element  = :0
 *******
 VectorX 0.999454 VectY 0.0330373 VectZ 0 norm 1
Element  = :1
 *******
 VectorX 0.999454 VectY 0.0330373 VectZ 0 norm 1

**Reply by GiudGiud** [reply]
Ah i remember now why we wanted stateful material properties. 

So once you have that random field on the first time step, you ll want to retrieve it in your code as the “old” material property

**Reply by selarem** [reply]
Thank you for your answers and time. 
I can not believe it needs that  much time to properly implement a very common feature.
Thus,  I am going to mimic random behavior by using trigonometric functions with high frequencies.

**Reply by selarem** [reply]
The easiest way to solve the problem is  to make random at initQpStatefulProperties and then retrieve it as "old" material property.
regards,

**Comment by ghost** [comment]
> Hello
> 
> You ll want to use the RandomIC on a variable then use that for the random number If it s not quite what you want and you cant find a conversion from the variable to what you need you can make a new object in the category you need and use the RandomInterface
> 
> Guillaume

Hello,

We're still trying to input random vector components using ParsedVectorFunction (we want this vector to remain unchanged during the entire computation, for each qp). We're trying to use RandomIC in the input file as follows : 

```
[ICs]
  [vectyy]
    type = RandomIC
    min = 0
    max = 1
    variable = vecty
  []
  [vectxx]
    type = RandomIC
    min = 0
    max = 1
    variable = vectx
  []
[]
```
To use in the Function : 
```

[Functions]
  [vectorexp]
    type = ParsedVectorFunction
    expression_x = 'vectx + x'
    expression_y = 'vecty + y'
    expression_z = '0'
  [../]
[]
```
We have defined the AuxVariables : 

```
[AuxVariables]
  [./vectx]
    order = CONSTANT
    family = MONOMIAL
  [../]
[./vecty]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]
```

But we get the error message l**ibMesh terminating:
ERROR: FunctionParser is unable to parse expression: vectx + x**
So how can we fix this ? Thank you.

Regards,

Nour

**Reply by GiudGiud** [reply]
hello

You cannot use ParsedVectorFunction with variables. 
Variables are semantically not known at the beginning of a simulation, while a Function should mostly be a known quantity at the beginning of the simulation./

Use this instead to build an auxvariable with the expression desired
https://mooseframework.inl.gov/source/auxkernels/ParsedVectorAux.html

**Comment by selarem** [comment]
We want random !

**Reply by GiudGiud** [reply]
You can do randomIC on an auxvariable then use that randomly populated field later in a ParsedVectorAux, that s not a problem

**Reply by GiudGiud** [reply]
I guess you are trying to convert the auxvariable to a material property next? For that you can use the ParsedMaterial with that randomIC-initialzed auxvariable

**Reply by ghost** [reply]
Thank you for your reply.
We have followed your suggestion and defined our AuxVariables:

[vectx]
  family = MONOMIAL
[]
[vecty]
   family = MONOMIAL
[]
 [parsed]
family = MONOMIAL_VEC
  []


then we expressed the parsed variable in the AuxKernels as:

[parsed]
    type = ParsedVectorAux
    variable = parsed
    expression_x = ' vectx'
    expression_y = ' vecty'
   coupled_variables = 'vectx vecty'
   use_xyzt = true
[]

and  IC as:

[ICs]
  [ppwater]
    type = RandomIC
    variable = vectx
    min = 0
    max = 10.0
  []
 [ICvecty]
    type = RandomIC
    variable = vecty
    min = 0
    max = 10.0
  []
[]

and all this work fine.



But our Material models needs to read a vectorValue property 
`_cp2(getMaterialProperty <RealVectorValue> ("MyDirection")),`


So in the Material block we wrote :

[./damage]
    type = MyDirectionalFracture
   MyDirection = parsed    
  [../]

Since with ParsedMaterial, we could not enter a vector expression.
We have also tried with ParsedVectorFunction again and both do not work.

regards,

**Reply by GiudGiud** [reply]
I don’t think there s a ParsedVectorMaterial you could use. 
You d have to make it. 

Is this your code? Can you switch from using a vector material property to specifying each of the components? Then you could take a ParsedMaterial

**Reply by ghost** [reply]
Do you have a clear idea about what we should do to solve this problem or are you  trying to answer our questions individually ?
 We are looking for a solution and spent a lot of  time trying.
We want to be able to initiate   random vector components. One vector for each qp that does not change during the computation. It should not take two months of back and forth for us to still not be on the same page regarding this question.

Regarding your last suggestion of specifying each component, we've done that in the .C file as :

```
_cp0(getMaterialProperty<Real>("vectx")),
   _cp1(getMaterialProperty<Real>("vecty")),
   _cp2(getMaterialProperty<Real>("vectz")),

```

in the input parameter properties. And then in the input file we defined the Material blocks:

```
[./vectx1111]
    type = ParsedMaterial
    expression = '2*vectx'
    property_name = vectx
    material_property_names = 'vectx'
  [../]
    [./vecty22222]
    type = ParsedMaterial
    expression = 55
    property_name = vecty
    material_property_names = 'vecty'
  [../]
[./vectz1111]
    type = ParsedMaterial
    expression = '0.0'
    property_name = vectz
    material_property_names = 'vectz'
  [../]
```
With IC
```
[ICs]
  [vectyy]
    type = RandomIC
    min = -1000
    max = 1000
    seed = 100
    variable = vecty
  []
  [vectxx]
    type = RandomIC
    min = -1000
    seed = 1000
    max = 1000
    variable = vectx
  []
[]
```
And AuxVariables 

```
[vectx]
family = MONOMIAL
[]
[vecty]
family = MONOMIAL
[]
```

By printing out the vector coordinates from the C file we get 0, 55, 0 respectively for vectx vecty vectz. 
How can we proceed from this?

**Reply by GiudGiud** [reply]
this is wrong
```
[./vectx1111]
    type = ParsedMaterial
    expression = '2*vectx'
    property_name = vectx
    material_property_names = 'vectx'
  [../]
    [./vecty22222]
    type = ParsedMaterial
    expression = 55
    property_name = vecty
    material_property_names = 'vecty'
  [../]
[./vectz1111]
    type = ParsedMaterial
    expression = '0.0'
    property_name = vectz
    material_property_names = 'vectz'
  [../]
```

it should be
```
[./vectx1111]
    type = ParsedMaterial
    expression = '2*vectx'
    property_name = vectx
    coupled_variables = 'vectx'
  [../]
    [./vecty22222]
    type = ParsedMaterial
    expression = 55
    property_name = vecty
    coupled_variables = 'vecty'
  [../]
[./vectz1111]
    type = ParsedMaterial
    expression = '0.0'
    property_name = vectz
    coupled_variables = 'vectz'
  [../]
```

vectxyz as an input is not a material property it s an auxiliary variable.
